a10networks/a10-octavia
# Add hierarchical multitenancy support for hardware devices

## Description
Provides for support for deployments using hierarchical multitenancy. For more information in the Openstack feature itself see: https://www.openstack.org/videos/summits/austin-2016/flat-no-more-hierarchical-multitenancy-and-projects-acting-as-domains-in-openstack

In this deployment model, projects will be mapped to L3V partitions based upon information provided in the `rack_device` config section.

Should the `use_parent_partition` config setting be set to true, then configuration bound for the child L3V partition will occur in the parent L3V partition.

## Limitations
Please note that if you are using `use_parent_partition` then the parent and the child partition should exist on the same Thunder device.

## Jira Ticket
Epic Link: [STACK-486](https://a10networks.atlassian.net/browse/STACK-486)

[STACK-1138](https://a10networks.atlassian.net/browse/STACK-1138)
[STACK-1062](https://a10networks.atlassian.net/browse/STACK-1062)
[STACK-1068](https://a10networks.atlassian.net/browse/STACK-1068)

## Technical Approach
For standard hierarchical multitenancy
- Added hierarchical_multitenancy to db and models.
- Added alembic revision script.
- Added enable_hierarchical_multitenancy config support in [a10_global] settings.
- Supported for enable_hierarchical_multitenancy in CreateRackVThunderEntry() and added revert().
- Refactored utils.py to add partition name length validation in config.
- Modified partition -> partition_name in db as well as configuration
- Added UTs for testing common/utils.py

For use_parent_partition enhancement
- Added use_parent_partition config setting
- Added logic to pull from parent project instead of current in various tasks

## Config Changes
<pre>
[a10_global]
<b>enable_hierarchical_multitenancy = True</b>
<b>use_parent_partition = True</b>

[rack_vthunder]

devices=[
 {
 "project_id": "eccaddd60f404477a7c32270267729d4",
 "username" : "admin",
 "password" : "password",
 "ip_address" : "10.43.12.137",
 "device_name" : "rack_1"
 },
 {
 "project_id": "6babdd52ba6143698170d0cad4392ca3",
 "username" : "admin",
 "password" : "password",
 "ip_address" : "10.43.12.137",
 "device_name" : "rack_1"
 }]
</pre>

## Test Cases
1. if OHM is disabled - the logic will not kick in 
2. if OHM is disabled, and use_parent_partition is enabled, warning should be printed and non-parent-partition will carry on 
3. if project's parent doesnt exists, current parition name is used - pass
4. if project's parent exists, and parent partition does not exists, a partition with parent project id is created 
5. if project's parent exists, and parent partition exists, parent partition name is used - pass
6. if parent project was created with custom name and then OHM and use_parent_partition was enabled, and tried to create child LB, it will not boot inside parent partition (we can solve it by reading parent object) - always use parent project id, do not honor custom name as OHM is enabled in config
7. if child vthunder is different than parent vthunder, a partition on child vthunder with parent project id is created and LB is placed in it.

## Manual Testing
Setup:
- Create project named corpA
- Create project named Engineering with corpA as parent
- Add Thunder device info to config (see above for example)

Test OHM:
- Set `enable_hierarchical_multitenancy = True` in config and create a load balancer from one of the projects
   - _Expected Result:_ Partition where lb is created should match first 14 char of project_id

Test parent partition:
- Set `use_parent_partition = True` in config and create loadbalancer from Engineering project
  - _Expected Result:_ Partition where lb is created should match first 14 char of corpA
